docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,235 @@
+
+##### BYO Resources
+
+BYO resources likely will not exist as ASO resources, so determining whether or not a given expected Azure resource exists becomes less straightforward using ASO. Initially, both BYO and CAPZ-managed resources will be created in ASO with ASO's `serviceoperator.azure.com/reconcile-policy: skip` label which tells ASO not to make any changes to the corresponding Azure resource, essentially assuming the resource is BYO. ASO will then perform a GET request against the Azure API to fill in the resource's `status`, which may indicate the resource does not exist with a status condition with a `reason: AzureResourceNotFound`. Then, if a future CAPZ reconciliation reveals the Azure resource does not exist, CAPZ will change it's assumption and make ASO "adopt" the resource by setting `serviceoperator.azure.com/reconcile-policy: manage` to allow changes made to the ASO resource to be reflected in Azure.
+
+#### API Changes
+
+Integration with ASO will not require any updates to CAPZ resources and CAPZ's CRDs will not require any updates. The only potentially user-affecting change is that the `status.longRunningOperationStates` on CAPZ resources will no longer be populated. This field is currently set for CAPZ's internal use to inform the `status.conditions` of the resource. With ASO, a CAPZ resource's `status.conditions` will now be derived from ASO resources' `status.conditions`.
+
+#### Installation
+
+ASO is a self-contained component published independently of CAPZ. When CAPZ is relying on ASO for critical functionality, CAPZ's installation process should also automatically install ASO.
+
+When installing CAPZ with `clusterctl init`, the Kubernetes resources defined by CAPZ's `infrastructure-components.yaml` published on the corresponding release are installed. ASO can also be [installed with a YAML manifest](https://azure.github.io/azure-service-operator/introduction/installing-from-yaml/) published for each of its releases describing its components that can be incorporated with CAPZ's `infrastructure-components.yaml`. This method installs ASO in the `azureserviceoperator-system` namespace and expects a Secret named `aso-controller-settings` to exist with Service Principal credentials. The easiest and most flexible way to customize that manifest further would likely be to generate a manifest from ASO's Helm chart with any necessary values overridden.
+
+Other than integrating with CAPZ's `infrastructure-components.yaml`, the `capz-controller-manager` container itself could assume responsibility for installing ASO by using the Helm Go library to install the ASO chart when it starts. Installing ASO this way allows CAPZ to hook in additional logic, such as to check if ASO is already installed in the cluster or to verify ASO is healthy after it's installed, which would not be possible if a static manifest was embedded in `infrastructure-components.yaml`.
```

Comment:
This would require a set of elevated permissions that many cluster operators (like myself) don't like to give custom-operators. There isn't an easy way with the RBAC so say CAPZ can only create/manage specific CRDs afaik

Reply:
That's a great point, I'll make sure to include that here.

Reply:
This _is_ a good point and it applies to ASO itself too, which is moving to an "the operator pod installs its own CRDs" approach rather than having the CRDs be in a Helm chart. See https://azure.github.io/azure-service-operator/design/adr-2023-02-helm-chart-size-limitations/. We can make those permissions optional as long as the CRDs are already installed (operator will probably still want LIST CRDs permissions to discover which are and are not but presumably that's less problematic than any Create/Update/Delete CRD option).

Reply:
From my perspective LIST is fine.

Reply:
Thanks @matthchr for that context. So I think no matter how we install ASO, we'd at least want to install the CRDs from infrastructure-components.yaml and configure ASO to assume they're already installed and not give it create/update/delete permissions for CRDs. I didn't see mention of the permissions aspect in the proposal or https://github.com/Azure/azure-service-operator/issues/1433, so @matthchr would it be best to open a new issue for that piece or add a comment on the existing issue?

Then putting the rest of the ASO components in infrastructure-components.yaml would complete the circle where CAPZ itself doesn't require any extra RBAC permissions to get ASO installed. @mtougeron Does that roughly match how you envision that working?

Reply:
> I didn't see mention of the permissions aspect in the proposal or https://github.com/Azure/azure-service-operator/issues/1433, so @matthchr would it be best to open a new issue for that piece or add a comment on the existing issue?

We had considered the permissions issue but had planned on just requiring them until a user requested an alternative. I've updated the [design document](https://github.com/Azure/azure-service-operator/pull/2765) to modify that plan. We'll just support user-managed CRDs out of the box.

Reply:
> That sounds good to me. As I understand this approach, the CRDs would get installed by the Cluster Operator (human) running `clusterctl init` using their credentials and the custom operators (CAPZ & ASO) would never have RBAC to do so. Is that correct?

@mtougeron Yep, that's my understanding.

> We had considered the permissions issue but had planned on just requiring them until a user requested an alternative. I've updated the [design document](https://github.com/Azure/azure-service-operator/pull/2765) to modify that plan. We'll just support user-managed CRDs out of the box.

Thanks for the input @matthchr, I'll update the proposal here to reflect that.